scripts/RunNodes.hs

```diff
@@ -0,0 +1,128 @@
+import Chainweb.Version
+import Control.Concurrent (threadDelay)
+import Control.Concurrent.Async
+import Control.Error.Util (note)
+import qualified Data.Attoparsec.Text as A
+import qualified Data.Text as T
+import Formatting
+import Options.Applicative
+import Shelly hiding (FilePath)
+import System.Directory (doesFileExist)
+
+---
+
+data Env = Env
+  { exe :: FilePath
+  , nodes :: Word8
+  , version :: ChainwebVersion
+  , logTarget :: Maybe Log
+  , logLevel :: T.Text
+  } deriving (Show)
+
+data Log = Path T.Text | ES T.Text Word deriving (Show)
+
+pEnv :: Parser Env
+pEnv = Env <$> pExe <*> pNodes <*> pVersion <*> optional pLogTarget <*> pLogLevel
+
+pExe :: Parser FilePath
+pExe = strOption
+  (long "exe" <> metavar "FILE"
+   <> help "Full path to a chainweb-node binary")
+
+pNodes :: Parser Word8
+pNodes = option auto
+  (long "nodes" <> metavar "COUNT" <> value 10
+   <> help "Number of Nodes to run (default: 10)")
+
+pVersion :: Parser ChainwebVersion
+pVersion = option cver
+  (long "version" <> metavar "VERSION"
+   <> value (TestWithTime petersonChainGraph)
+   <> help "Chainweb Version to run the Nodes with (default: testWithTime)")
+  where
+    cver :: ReadM ChainwebVersion
+    cver = eitherReader $ \s ->
+        note "Illegal ChainwebVersion" . chainwebVersionFromText $ T.pack s
+
+pLogTarget :: Parser Log
+pLogTarget = option logPath
+  (long "log-target" <> metavar "local:PATH|es:ELASTICSEARCH_HOST:PORT"
+   <> help "Location to log to (default: STDOUT)")
+  where
+    logPath :: ReadM Log
+    logPath = eitherReader (A.parseOnly (esPath <|> plainPath) . T.pack)
+
+    esPath :: A.Parser Log
+    esPath = ES <$> host <*> port
+      where
+        host = A.string "es:" *> A.takeWhile1 (/= ':') <* A.char ':'
+        port = A.decimal
+
+    plainPath :: A.Parser Log
+    plainPath = Path <$> (A.string "local:" *> A.takeText)
+
+pLogLevel :: Parser T.Text
+pLogLevel = strOption
+  (long "log-level" <> metavar "info|warn|error|debug" <> value "info"
+   <> help "default: info")
+
+runNode :: Word8 -> Maybe T.Text -> Env -> IO ()
+runNode nid config (Env e ns v lt ll) = shelly $ run_ (fromText $ T.pack e) ops
+  where
+    logging :: Log -> [T.Text]
+    logging l@(ES _ _) =
+      [ sformat ("--telemetry-log-handle=" % stext) $ logText l
+      , sformat ("--log-handle=" % stext) $ logText l
+      ]
+    logging l@(Path _) =
+      [ sformat ("--telemetry-log-handle=" % stext % "/telemetry.node" % int % ".log") (logText l) nid
+      , sformat ("--log-handle=" % stext % "/node" % int % ".log") (logText l) nid
+      ]
+
+    logText :: Log -> T.Text
+    logText (ES h p) = sformat ("es:" % stext % ":" % int) h p
+    logText (Path p) = sformat ("file:" % stext) p
+
+    ops :: [T.Text]
+    ops = [ "--hostname=127.0.0.1"
+          , sformat ("--node-id=" % int) nid
+          , sformat ("--test-miners=" % int) ns
+          , sformat ("--chainweb-version=" % stext) $ plainVersionFromText v
```

Comment:
I am somewhat concerned about stripping the graph information from the version here. We'll soon start working on hard-forks. So, will probably have to deal more often with different graphs. It looks like this code will reset the version to the default graph, even if the user asked for a different graph.

Reply:
I had assumed that `--chainweb-version` would not accept a version string that had a graph indicated. Let me double-check that.

Reply:
Fixed.